Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,239 @@
+
+                        group.cancelAll()
+                        throw ServiceRunnerError.serviceFinishedUnexpectedly()
+                    } else {
+                        // This service finished early but we expected it.
+                        // So we are just going to wait for the next service
+                        self.logger.debug(
+                            "Service finished",
+                            metadata: [
+                                self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                            ]
+                        )
+
+                        // We have to keep track of how many services finished to make sure
+                        // to stop when only the signal child task is left
+                        finishedServiceCount += 1
+                        if finishedServiceCount == self.services.count {
+                            // Every service finished. We can cancel the signal handling now
+                            // and return
+                            group.cancelAll()
+                            return
+                        } else {
+                            // There are still running services that we have to wait for
+                            continue
+                        }
+                    }
+
+                case .serviceThrew(let service, let error):
+                    // One of the servers threw an error. We have to cancel everything else now.
+                    self.logger.error(
+                        "Service threw error. Cancelling all other services now",
+                        metadata: [
+                            self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                            self.configuration.loggingConfiguration.errorKey: "\(error)",
+                        ]
+                    )
+                    group.cancelAll()
+
+                    throw error
+
+                case .signalCaught(let unixSignal):
+                    // We got a signal. Let's initiate graceful shutdown in reverse order than we started the
+                    // services. This allows the users to declare a hierarchy with the order they passed
+                    // the services.
+                    self.logger.info(
+                        "Signal caught",
+                        metadata: [
+                            self.configuration.loggingConfiguration.signalKey: "\(unixSignal)",
+                        ]
+                    )
+
+                    for service in self.services.reversed() {
+                        self.logger.debug(
+                            "Shutting down service",
+                            metadata: [
+                                self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                            ]
+                        )
+
+                        do {
+                            try await service.shutdownGracefully()
```

Comment:
With the default implementation that we provide for Service this would hang forever:

```swift
extension Service {
     public var isLongRunning: Bool {
         // By default all services are treated as long running services.
         true
     }

     public func run() async throws {
         // We are just going to suspend here until we get cancelled.
         try await CancellableContinuation().run()
     }

     public func shutdownGracefully() async throws {
         // no-op
     }
}
```

Thinking more about this, I have the feeling that we work around a limitation in TaskGroup with the `shutdownGracefully`. The problem is that taskGroup does not allow us to cancel child tasks individually. However I think we should not leak this deficiency in task group into our API with `shutdownGracefully`. Instead for now we should use unstructured Tasks inside the ServiceRunner and add a big feature request for the Swift team.

Reply:
This is not what we are trying to solve. See the comment above w.r.t. to why we need `shutdownGracefully`

Reply:
Fine, but then we should not have a default implementation for `shutdownGracefully` that does nothing :)

Reply:
I agree that the default implementations are tricky. Maybe we shouldn't have any at all?

Reply:
> Maybe we shouldn't have any at all?

+1